acra-censor/common/logging_logic.go

```diff
@@ -0,0 +1,285 @@
+	if err != nil {
+		queryWriter.logger.WithError(ErrCantOpenFileError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't dump queries to file")
+		return err
+	}
+	// write all queries
+	return queryWriter.appendQueries(queryWriter.Queries, f)
+}
+
+// Release dumps all Captured queries to file and resets Captured queries list.
+func (queryWriter *QueryWriter) Release() {
+	queryWriter.signalBackgroundExit <- true
+	queryWriter.reset()
+}
+
+func (queryWriter *QueryWriter) reset() {
+	queryWriter.Queries = nil
+	queryWriter.BufferedQueries = nil
+}
+
+func (queryWriter *QueryWriter) readAllQueriesFromFile() error {
+	q, err := readQueries(queryWriter.filePath)
+	if err != nil {
+		queryWriter.logger.WithError(ErrCantReadQueriesFromFileError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't read queries from file")
+		return err
+	}
+	// read existing queries from file
+	queryWriter.Queries = q
+	return nil
+}
+
+func readQueries(filePath string) ([]*QueryInfo, error) {
+	bufferBytes, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+	var queries []*QueryInfo
+	if len(bufferBytes) != 0 {
+		for _, line := range bytes.Split(bufferBytes, []byte{'\n'}) {
+			if len(line) == 0 {
+				continue
+			}
+			var oneQuery QueryInfo
+			if err = json.Unmarshal(line, &oneQuery); err != nil {
+				return nil, err
+			}
+			queries = append(queries, &oneQuery)
+		}
+	}
+	return queries, nil
+}
+
+func (queryWriter *QueryWriter) dumpBufferedQueriesToFile(openedFile *os.File) error {
+	// nothing to dump
+	if len(queryWriter.BufferedQueries) == 0 {
+		return nil
+	}
+	err := queryWriter.appendQueries(queryWriter.BufferedQueries, openedFile)
+	if err != nil {
+		return err
+	}
+	// clean buffered queries only after successful write
+	queryWriter.BufferedQueries = nil
+	return nil
+}
+
+func (queryWriter *QueryWriter) appendQueries(queries []*QueryInfo, openedFile *os.File) error {
+	if len(queries) == 0 {
+		return nil
+	}
+	lines := queryWriter.serializeQueries(queries)
+	if _, err := openedFile.Write(lines); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (queryWriter *QueryWriter) serializeQueries(queries []*QueryInfo) []byte {
+	var linesToAppend []byte
+	var tempQueryInfo = &QueryInfo{}
+	for _, queryInfo := range queries {
+		tempQueryInfo.RawQuery = queryInfo.RawQuery
+		tempQueryInfo.IsForbidden = queryInfo.IsForbidden
+		jsonQueryInfo, err := json.Marshal(tempQueryInfo)
+		if err != nil {
+			queryWriter.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorQuerySerializeError).Errorln("Can't serialize stored queries")
+		}
+		if len(jsonQueryInfo) > 0 {
+			jsonQueryInfo = append(jsonQueryInfo, '\n')
+			linesToAppend = append(linesToAppend, jsonQueryInfo...)
+		}
+	}
+	return linesToAppend
+}
+
+func (queryWriter *QueryWriter) finishAndCloseFile(openedFile *os.File) error {
```

Comment:
https://golang.org/pkg/io/#Closer enough here. interfaces better to use instead specific types

Reply:
Can you please, explain in more details